metabase/metabase
# Support all date buckets on waterfall charts

Closes https://github.com/metabase/metabase/issues/37320
Closes https://github.com/metabase/metabase/issues/37305

### Description

This PR adds timeseries unit inferrence from data even for native queries which lets us compute the waterfall total x-axis value correctly. In addition to that, it lets us group timeline events by unit on native queries.

### How to verify

- Create a timeseries waterfall chart in GUI
- Test it on all datetime units and ensure it looks good
- Create a timeseries waterfall chart using an SQL query
- Ensure it correctly infers bucket from the data and bar widths are sensible

Timeline events
Ensure timeline events on native queries are grouped correctly. For instance if data contains monthly granularity and you have two events Jan 1st and Jan 31st, they should be grouped under the tick that represents January.

Sample query (converted from GUI)
```
SELECT
  DATE_TRUNC('month', "PUBLIC"."ORDERS"."CREATED_AT") AS "CREATED_AT",
  COUNT(*) AS "count",
  AVG("PUBLIC"."ORDERS"."TOTAL") AS "avg"
FROM
  "PUBLIC"."ORDERS"
WHERE
  (
    "PUBLIC"."ORDERS"."CREATED_AT" >= timestamp with time zone '2023-04-01 00:00:00.000-03:00'
  )
 
   AND (
    "PUBLIC"."ORDERS"."CREATED_AT" < timestamp with time zone '2023-10-01 00:00:00.000-03:00'
  )
GROUP BY
  DATE_TRUNC('month', "PUBLIC"."ORDERS"."CREATED_AT")
ORDER BY
  DATE_TRUNC('month', "PUBLIC"."ORDERS"."CREATED_AT") ASC
```
Add timeline events on 1 Aug 2023 and 31 Aug 2023 and ensure they are grouped (or check the screenshots below).

### Demo
Github does not upload my images, so I uploaded images to a random image uploading service.

Before
https://pasteboard.co/0aVN3tYHQhzW.png

After
https://pasteboard.co/UvKhrRkhGfFq.png

**Timeline events**

Before
https://pasteboard.co/ppI1XHvRCIgl.png

After
https://pasteboard.co/UGYlvfDgLpJb.png

### Checklist
